src/MongoDB.Driver.Core/Core/Compression/Zstd/ZstandardStream.cs

```diff
@@ -0,0 +1,380 @@
+            }
+            return compressionMode;
+        }
+
+        // nested types
+        private class StreamReadHelper
+        {
+            private readonly byte[] _compressedDataBuffer;
+            private readonly Stream _compressedStream;
+            private readonly ReadingState _readingState;
+
+            public StreamReadHelper(
+                Stream compressedStream,
+                byte[] compressedBuffer)
+            {
+                _compressedDataBuffer = Ensure.IsNotNull(compressedBuffer, nameof(compressedBuffer));
+                _compressedStream = Ensure.IsNotNull(compressedStream, nameof(compressedStream));
+                _readingState = new ReadingState();
+            }
+
+            public BufferInfo CompressedBufferInfo => new BufferInfo(_compressedDataBuffer, _readingState.DataPosition);
+
+            // public methods
+            public int ReadCompressedStreamToBufferIfAvailable(
+                int recommendedCompressedSize,
+                out int remainingSize)
+            {
+                remainingSize = _readingState.LastReadDataSize - _readingState.DataPosition;
+
+                if (remainingSize <= 0 && !_readingState.IsDataDepleted && !_readingState.SkipDataReading)
+                {
+                    var readSize = _compressedStream.Read(_compressedDataBuffer, 0, recommendedCompressedSize);
+                    UpdateStateAfterReading(readSize, out remainingSize);
+                }
+
+                return _readingState.DataPosition;
+            }
+
+            public bool TryPrepareDataForNextAttempt(int compressedBufferPosition, int uncompressedBufferPosition)
+            {
+                if (uncompressedBufferPosition == 0) // 0 - when a frame is completely decoded and fully flushed
+                {
+                    // the internal buffer is depleted, we're either done
+                    if (_readingState.IsDataDepleted)
+                    {
+                        return false;
+                    }
+
+                    // or we need more bytes
+                    _readingState.SkipDataReading = false;
+                }
+
+                // 1.calculate progress in compressed(input) buffer
+                // 2. save the data position for next Read calls
+                _readingState.DataPosition += compressedBufferPosition;
+                return true;
+            }
+
+            // private methods
+            private void UpdateStateAfterReading(int readSize, out int remainingSize)
+            {
+                _readingState.LastReadDataSize = readSize;
+                _readingState.IsDataDepleted = readSize <= 0;
+                _readingState.DataPosition = 0;
+
+                // skip _compressedStream.Read until the internal buffer is depleted
+                // avoids a Read timeout for applications that know the exact number of bytes in the _compressedStream
```

Comment:
Question: Does the word "applications" here refer to the applications that use the driver?